### PR TITLE
Custom RBF and Matern Functions

### PR DESCRIPTION
**Pull request by KeAWang**
This pull request tries to reduce the memory requirements of a RBFKernel and a MaternKernel by introducing a custom backward function when the lengthscales require grad.
- [x] Implement RBFCovariance Function. Fallback to RBFKernel when using ARD or when x1, x2 require grad
- [x] Implement MaternCovariance Function. Fallback to MaternKernel when using ARD or when x1, x2 require grad
- [x] Add unittests for forward and backward

ARD is not yet implemented because I'm not sure if there's a better way to do it than to create a distance matrix for each dimension (resulting in a 3D Tensor) since we need gradients of a 2D Tensor with respect to each of the lengthscales.

### CODE REVIEW

**Comment by Balandat** [comment]
Haven't really looked into this, just wanted to make a side comment here: In light of #492, we'll want to start support arbitrary batch shapes also in the kernels - so I guess this is a call to be cognizant of that, and to make sure the code will be easily adaptable to these coming changes.

**Comment by KeAWang** [comment]
Fixed. Thanks for pointing that out. Also, I noticed that `MaternKernel.forward` might not actually be arbitrary batch size compatible because how it computes the mean will mess up the broadcasting. I added a commit that fixes that.

I'm worried that there might be a lot of other code that assume there's only one batch dimension but we don't know about them because our tests all assume there's only batch dimension. @jacobrgardner @gpleiss ?

**Comment by gpleiss** [comment]
@KeAWang #492

**Comment by gpleiss** [comment]
I'd rather not merge in this PR now - I think we should test against this PR for a while to see if we can make further memory gains.

**Comment by Balandat** [comment]
> I'm worried that there might be a lot of other code that assume there's only one batch dimension but we don't know about them because our tests all assume there's only batch dimension.

Yes, I think outside of the LazyTensor refactor most other code assumes a single batch dimension, this will all need to be changed. Most of the hard work should be done in @gpleiss's refactor though.

**Comment by KeAWang** [comment]
Here's a quick memory usage benchmark from the output of `torch.cuda.max_memory_allocated` on 10000 1 dimensional training points:

Using RBFKernel on master:
   after forward: 1.2 GB
   after backward: 2.1 GB
Using RBFCovariance:
   after forward: 1.2 GB
   after backward: 1.6 GB
Using MaternKernel (nu=1/2) on master:
   after forward: 2.4 GB
   after backward: 2.8 GB
Using MaternFunction (nu=1/2):
   after forward: 1.2 GB
   after backward: 1.6 GB
Using MaternKernel (nu=3/2) on master:
   after forward: 2.8 GB
   after backward: 3.6 GB
Using MaternFunction (nu=3/2):
   after forward: 1.2 GB
   after backward: 1.6 GB
Using MaternKernel (nu=5/2) on master:
   after forward: 3.2 GB
   after backward: 4.4 GB
Using MaternCovariance (nu=5/2):
   forward: 1.6 GB
   backward: 1.6 GB

File: README.md [hunk 1]
```diff
@@ -92,7 +92,7 @@ GPyTorch is primarily maintained by:
 
 <img width="300" src=https://brand.cornell.edu/assets/images/downloads/logos/cornell_logo_simple/cornell_logo_simple.svg alt="Cornell Logo" />
 
-We would like to thank our other contributors including (but not limited to) Eytan Bakshy, David Arbour, Ruihan Wu, Bram Wallace, Sam Stanton, and Jared Frank.
+We would like to thank our other contributors including (but not limited to)  David Arbour, Eytan Bakshy, Jared Frank, Sam Stanton, Bram Wallace, Ke Alexaner Wang, Ruihan Wu.
```
Comment (jacobrgardner) [inline README.md]: @KeAWang At least spell your name right :-)

**Comment by jacobrgardner** [comment]
@KeAWang Can you make sure to keep this branch up to date with any changes you make to the kernels on the `implicit_kernel` branch? It'll probably be cleaner to merge those features separately if possible. If not, we can do that whole branch at once I suppose.

File: README.md (hunk 1 above)
Comment (KeAWang) [reply README.md]: Oops! This is embarrassing...

**Comment by KeAWang** [comment]
Yep I've been pushing to this remote branch and then merging into the `implicit_kernel` branch

**Comment by KeAWang** [comment]
This should be ready for review now

File: gpytorch/kernels/rbf_kernel.py [hunk 2]
```diff
@@ -94,7 +95,18 @@ def __init__(
...
+        if (
+            x1.requires_grad
+            or x2.requires_grad
+            or (self.ard_num_dims is not None and self.ard_num_dims > 1)
+            or diag
+        ):
+            x1_ = x1.div(self.lengthscale)
+            x2_ = x2.div(self.lengthscale)
+            diff = self._covar_dist(x1_, x2_, square_dist=True, diag=diag, **params)
+            return postprocess_rbf(diff)
```
Comment (jacobrgardner) [inline gpytorch/kernels/rbf_kernel.py]: @KeAWang Why does postprocess get called out of the JIT function now? The purpose of having this method at all was  was so that the full computation would be done inside a JIT script.

If we're not going to have postprocess scripts and only do the distance code in a JIT script, then the `dist_postprocess_func` arg needs to be removed everywhere and the function inlined here.

**Comment by jacobrgardner** [comment]
This looks good to me other than my comment above about the distance postprocessing functions.

**Comment by KeAWang** [comment]
Hm I'm not sure why this was there. It might have been left over from my earlier implementation of RBFCovariance. I reverted this change but now `test_fantasy_updates` mysteriously fails...

**Comment by KeAWang** [comment]
Fixed.
